File: solutions/150.evaluate-reverse-polish-notation.js
/*
 * @lc app=leetcode id=150 lang=javascript
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function (tokens) {
  const stack = [];
  tokens.forEach((t) => {
    if (/[0-9]/.test(t)) {
      stack.push(parseInt(t));
    } else {
      const b = stack.pop();
      const a = stack.pop();
      switch (t) {
        case "+":
          stack.push(b + a);
          break;
        case "-":
          stack.push(a - b);
          break;
        case "*":
          stack.push(a * b);
          break;
        case "/":
          stack.push(parseInt(a / b));
          break;
      }
    }
  });
  return stack[0];
};
// @lc code=end

File: solutions/150.evaluate-reverse-polish-notation.py
from typing import List


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        from collections import deque as stack

        stack = []
        ops = {
            '+': lambda a,b: a+b,
            '-': lambda a,b: a-b,
            '*': lambda a,b: a*b,
            '/': lambda a,b: int(a/b)
        }
        for t in tokens:
            if t in ops:
                b, a = stack.pop(), stack.pop()
                stack.append(ops[t](a,b))
            else:
                stack.append(int(t))
        return stack.pop()
